Add AWS auth method for cel and httpjson inputs

moukoublen opened the pull request:
## Proposed commit message
Add a new auth method for cel and httpjson that wraps the aws v4 signer.

As discussed [here](https://github.com/elastic/security-team/issues/13112) this PR introduces an new auth method that can be used in cel and httpjson integrations. 

This AWS auth method is implemented by creating an HTTP transport that wraps the transport of the client, uses a configuration, AWS credentials, and v4 signer, and signs the request just before passing it to the next (actual) HTTP transport. 


example of cel and httpjson configs that uses the aws auth:

```yaml
### cel
config_version: 2
interval: {{interval}}
auth.aws:
  access_key_id: {{access_key_id}}
  secret_access_key: {{secret_access_key}}
```

```yaml
### cel
config_version: 2
interval: {{interval}}
auth.aws:
  role_arn: {{role_arn}}
  external_id: {{external_id}}
```


```yaml
### httpjson
config_version: "2"
interval: {{interval}}
auth.aws:
  role_arn: {{role_arn}}
  external_id: {{external_id}}
```

The auth method config can be initialized with static credentials as well as assume role arn and every other available config in [ConfigAWS](https://github.com/moukoublen/beats/blob/input_aws_auth_method/x-pack/libbeat/common/aws/credentials.go#L35).

The service name and default region can be configured or inferred from the URL of the request. 


When using this method, the integration (cel or httpjson file) does not need to re-implement the aws sing process inside the file, while at the same time having the singer as http.Transport, offers some advantages (described [here](https://github.com/elastic/security-team/issues/13112#issuecomment-3139188187); like enabling an integration using all available aws auth methods based on user config with a single cel/httpjson file) and will enable the cloud connectors project to be expanded to other aws integrations (cloud connector specific changes will be introduced in a later PR). 


## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent-changelog-tool/blob/main/docs/usage.md).

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Beats.
-->

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
-

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
Pinging @elastic/obs-ds-hosted-services (Team:obs-ds-hosted-services)

elasticmachine commented:
Pinging @elastic/security-service-integrations (Team:Security-Service Integrations)

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

leehinman commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -297,7 +298,21 @@ func newHTTPClient(ctx context.Context, authCfg *authConfig, requestCfg *request
 		err    error
 	)
-	if authCfg.OAuth2.isEnabled() {
+	if authCfg.AWS.IsEnabled() {
```
nit.  Since we now have `else if` any chance I can convince you to switch to `case` (pun intended).

leehinman reviewed (changes requested):
needs a changelog fragment, but code LGTM

efd6 commented on changelog/fragments/1761377900-input-auth-method-aws.yaml:
```diff
@@ -0,0 +1,45 @@
+
+# REQUIRED for all kinds
+# Change summary; a 80ish characters long description of the change.
+summary: Add AWS auth method for cel and httpjson.
```
```suggestion
summary: Add AWS auth method for CEL and HTTP JSON inputs.
```
The text here is user-facing, so the text should match the way that it is present to them in the rest of the documentation.

efd6 commented on x-pack/filebeat/input/cel/input.go:
```diff
@@ -176,6 +177,16 @@ func (i input) run(env v2.Context, src *source, cursor map[string]interface{}, p
 	}
+
+	if cfg.Auth.AWS.IsEnabled() {
+		tr, err := aws.InitializeSingerTransport(*cfg.Auth.AWS, log, client.Transport)
```
```suggestion
		tr, err := aws.InitializeSignerTransport(*cfg.Auth.AWS, log, client.Transport)
```

efd6 commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -297,7 +298,22 @@ func newHTTPClient(ctx context.Context, authCfg *authConfig, requestCfg *request
+		}
+
+		log.Debugf("aws auth method: creating signer for region: %s and service: %s", authCfg.AWS.DefaultRegion, authCfg.AWS.ServiceName)
+		tr, err := aws.InitializeSingerTransport(*authCfg.AWS, log, client.Transport)
```
```suggestion
		tr, err := aws.InitializeSignerTransport(*authCfg.AWS, log, client.Transport)
```

efd6 commented on x-pack/filebeat/input/cel/input.go:
```diff
@@ -176,6 +177,16 @@ func (i input) run(env v2.Context, src *source, cursor map[string]interface{}, p
+	if cfg.Auth.AWS.IsEnabled() {
+		tr, err := aws.InitializeSingerTransport(*cfg.Auth.AWS, log, client.Transport)
+		if err != nil {
+			log.Errorw("aws auth method: initialize of aws config failed for aws signer", "error", err)
```
```suggestion
			log.Errorw("initialize of aws config failed for aws signer", "error", err)
```
There's enough text to find this without the colon prefix.

efd6 commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -297,7 +298,22 @@ func newHTTPClient(ctx context.Context, authCfg *authConfig, requestCfg *request
+	case authCfg.AWS.IsEnabled():
+		client, err = newNetHTTPClient(ctx, requestCfg, log, reg)
+		if err != nil {
+			log.Errorw("aws auth method: creation of initial http client failed", "error", err)
```
```suggestion
			log.Errorw("creation of initial http client failed", "error", err)
```

efd6 commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -297,7 +298,22 @@ func newHTTPClient(ctx context.Context, authCfg *authConfig, requestCfg *request
+		log.Debugf("aws auth method: creating signer for region: %s and service: %s", authCfg.AWS.DefaultRegion, authCfg.AWS.ServiceName)
+		tr, err := aws.InitializeSingerTransport(*authCfg.AWS, log, client.Transport)
+		if err != nil {
+			log.Errorw("aws auth method: initialize of aws config failed for aws signer", "error", err)
```
```suggestion
			log.Errorw("initialize of aws config failed for aws signer", "error", err)
```

efd6 commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -297,7 +298,22 @@ func newHTTPClient(ctx context.Context, authCfg *authConfig, requestCfg *request
+			return nil, err
+		}
+
+		log.Debugf("aws auth method: creating signer for region: %s and service: %s", authCfg.AWS.DefaultRegion, authCfg.AWS.ServiceName)
```
```suggestion
		log.Debugf("creating signer for region: %s and service: %s", authCfg.AWS.DefaultRegion, authCfg.AWS.ServiceName)
```

efd6 commented on x-pack/libbeat/common/aws/signer.go:
```diff
@@ -0,0 +1,195 @@
+	// resolve service name and region (if they are not configured)
+	serviceName, region, err := st.getServiceAndRegion(req)
+	if err != nil {
+		return nil, fmt.Errorf("error while getting service name and region: [%w]", err)
```
```suggestion
		return nil, fmt.Errorf("error while getting service name and region: %w", err)
```
throughout (this is not conventional error syntax for Go)

efd6 commented on x-pack/libbeat/common/aws/signer.go:
```diff
@@ -0,0 +1,195 @@
+	// with *bytes.Buffer, *bytes.Reader or *strings.Reader as body, which gets GetBody initialized.
+	// httpjson: (x-pack/filebeat/input/httpjson/request.go newHTTPRequest)
+	// cel: mito repo (lib/http.go)
+	// We cover the edge case here by reading and coping the body manually.
```
```suggestion
	// We cover the edge case here by reading and copying the body manually.
```

moukoublen commented:
@efd6 Thanks for taking a look. I addressed all the comments apart from the package goddoc one (answered inline). I didn't resolve the comments in case you want to do that. 

> I'm concerned about the location of the shared code; because beats does not use valid Go module versioning using this in the mito developer tool will be difficult.

Do you think this code should be moved to mito? 

My only concern is that the current implementation relies on the already existing AWS config initialization that exists in the `aws` package and is used in both CEL and HTTP JSON inputs and mito is CEL specific if I am not mistaken. And moving forward, the flow of cloud connectors will be implemented inside the `InitializeAWSConfig` (which is the goal to use the cloud connectors flow through the AWS auth method).  

Perhaps a stripped-down version of the signer without the `SignerInputConfig` can be moved to mito and then used by the `InitializeSignerTransport`? But will that help with what you described? I am not familiar with the mito developer tool and how it works so far.

moukoublen commented on x-pack/filebeat/input/cel/input.go:
```diff
@@ -176,6 +177,16 @@ func (i input) run(env v2.Context, src *source, cursor map[string]interface{}, p
 		}
 	}
+
+	if cfg.Auth.AWS.IsEnabled() {
```
Nice catch. The signer must run "after" any other transports in order to be on the safe side. If any other transport (logs/tracer, metrics) adds any header, this must happen before the signer signs. 

I moved it to `newClient` on the top of the client initialization (so all other transports to wrap this one).

efd6 commented on x-pack/filebeat/input/cel/input.go:
```diff
@@ -814,6 +816,15 @@ func newClient(ctx context.Context, cfg config, log *logp.Logger, reg *monitorin
 		}
+	} else if cfg.Auth.AWS.IsEnabled() {
+		// this transport runs after the other ones (the other ones wrap this one); just to be on the safe side.
+		// If any of the other transports add any header, it must happen before the singing.
```
`s/signing/signing/`

efd6 commented:
One of the linter complaints is valid (the `%v` → `%w` change), the others are either due to weakness in errcheck's data flow analysis or outright incorrectness in the linter's approach.

olegsu commented:
@mergifyio backport 8.19

andrewkroh commented:
@Mergifyio backport 8.19
